tests/test_transform.py

```diff
@@ -119,6 +120,28 @@ def test_process_zipfile_failed_pdf(self, fake_decapitate):
         # clean the test directories
         clean_test_directories()
 
+
+    @patch.object(transform, 'decapitate_pdf_with_error_check')
+    def test_copy_pdf_to_output_dir_timed_out_pdf(self, fake_decapitate):
+        "tests of when pdf decapitaion reaches timeout"
+        # override the timeout value with 0 seconds to simulate a timeout
+        transform.PDF_DECAPITATE_TIMEOUT = 0
+        zipfile_name = os.path.join(TEST_DATA_PATH,
+                                    '18022_1_supp_mat_highwire_zip_268991_x75s4v.zip')
+        file_title_map = {
+            '18022_1_merged_1463214271.pdf': 'Merged PDF'
+        }
+        doi = '10.7554/eLife.12717'
+        current_zipfile = zipfile.ZipFile(zipfile_name, 'r')
+        with self.assertRaises(Exception) as context:
+            return_value = transform.copy_pdf_to_output_dir(file_title_map, None, doi,
+                                                            current_zipfile, POA_CONFIG)
+        current_zipfile.close()
+        # set the timed out value back again
```

Comment:
This cleanup should be part of a `tearDown` method, otherwise an uncaught error in `os.path.join` or other parts of this method won't execute it, leading to other tests being affected.
An Pythonic alternative would be a context manager created for the tests https://docs.python.org/2.7/library/contextlib.html#contextlib.contextmanager (or perhaps as part of the library, it can be useful to customize the timeout)

Reply:
I think in the case of this one test case where only one zip file is opened, it would be ok to nest two ``with`` statements to get around it and to satisfy closing the zip file - not perfect, but arguably acceptable here.

A better application of ``with`` and a ``contextmanager`` is probably in ``transform.process_zipfile()`` itself, which opens two zip files, reads from one, writes to the other, having no ``with`` statements at all. We could create a new ticket for that and improve it in a separate branch?

This library is not deployed yet, and it is probably good timing to identify improvements now before it does go into production. The old library we currently use hasn't had many runtime errors do the same operations, fortunately.